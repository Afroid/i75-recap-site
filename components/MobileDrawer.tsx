import { useState } from "react";
import Link from "next/link";
import { AnimatePresence, motion } from "framer-motion";
import { getRecaps } from "@/lib/getRecaps";
import { ChevronDown, ChevronUp } from "lucide-react";

const recapData = getRecaps();

interface MobileDrawerProps {
  isOpen: boolean;
  onClose: () => void;
}

export default function MobileDrawer({ isOpen, onClose }: MobileDrawerProps) {
  const [showOlderSeasons, setShowOlderSeasons] = useState(false);
  const [expandedYear, setExpandedYear] = useState<string | null>(null);

  const toggleOlderSeasons = () => setShowOlderSeasons((prev) => !prev);

  // Sort all of the seasons in descending order
  const sortedSeasons = Object.entries(recapData).sort(
    (a, b) => Number(b[0]) - Number(a[0])
  );

  const latestSeasons = sortedSeasons.slice(0, 3);
  const olderSeasons = sortedSeasons.slice(3);

  const renderSeasonLinks = (seasons: [string, number[]][]) =>
    seasons.map(([year, weeks]) => {
      const isExpanded = expandedYear === year;

      return (
        <div key={year} className="mb-2">
          <button
            onClick={() => setExpandedYear(isExpanded ? null : year)}
            className="
              flex justify-between items-center w-full
              rounded-md whitespace-nowrap text-gray-900
              ml-2
            "
          >
            {year} Season
            {isExpanded ? <ChevronUp className="w-4 h-4" /> : <ChevronDown className="w-4 h-4" />}
          </button>

          <AnimatePresence initial={false}>
          {isExpanded && (
            <motion.div
              initial={{ opacity: 0, height: 0 }}
              animate={{ opacity: 1, height: "auto" }}
              exit={{ opacity: 0, height: 0 }}
              transition={{ duration: 0.3, ease: "easeInOut" }}
              className="ml-5 overflow-hidden flex flex-col gap-1 mt-1"
            >
              {weeks.map((week) => (
                <Link
                  key={week}
                  href={`/recaps/${year}/week-${week}`}
                  onClick={onClose}
                  className="text-sm text-gray-700"
                >
                  Week {week}
                </Link>
              ))}
            </motion.div>
          )}
          </AnimatePresence>
        </div>
      );
    });

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.div
          className="fixed top-12 left-0 right-0 bottom-0 bg-black/50 z-[1000]"
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
          onClick={onClose}
        >
          <motion.div
            className="
              fixed top-12 left-0 w-60 h-[calc(100vh-3rem)]
              bg-gray-50
              pl-4 pr-6 py-2 z-[1001]
              flex flex-col gap-6"
            initial={{ x: "-100%" }}
            animate={{ x: 0 }}
            exit={{ x: "-100%" }}
            transition={{ type: "tween", duration: 0.4 }}
            onClick={(e) => e.stopPropagation()}
          >
            <div>
              <h2 className="font-bold text-lg text-green-600 mb-2 underline">
                Recaps
              </h2>

              {/* Latest Recaps */}
              {renderSeasonLinks(latestSeasons)}

              {showOlderSeasons && (
                <>
                  <div className="border-t border-gray-300 my-2" />
                  <h2 className="font-bold text-lg text-green-600 mb-2 underline">
                    Older Recaps
                  </h2>
                  {renderSeasonLinks(olderSeasons)}
                </>
              )}

              <div className="border-t border-gray-300 my-2" />

              <button
                onClick={toggleOlderSeasons}
                className="mb-2 text-md text-green-600"
              >
                {showOlderSeasons ? "Hide Older Seasons" : "View Older Seasons"}
              </button>

              <div className="mb-2">
                <Link
                  href="/recaps"
                  onClick={onClose}
                  className="text-md text-blue-600"
                >
                  View All Recaps
                </Link>
              </div>

              {/* Divider */}
              <div className="border-t border-gray-300" />
            </div>

            <div className="flex flex-col gap-4">
              <Link href="/" onClick={onClose} className="text-lg text-gray-800">
                Home
              </Link>
              <Link href="/about" onClick={onClose} className="text-lg text-gray-800">
                About
              </Link>
              <Link href="/contact" onClick={onClose} className="text-lg text-gray-800">
                Contact
              </Link>
            </div>
          </motion.div>
        </motion.div>
      )}
    </AnimatePresence>
  );
}
