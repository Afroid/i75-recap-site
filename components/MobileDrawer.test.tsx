import React, { HTMLAttributes } from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import MobileDrawer from "./MobileDrawer";

// Stub framer motion
jest.mock("framer-motion", () => ({
  AnimatePresence: ({ children }: { children: React.ReactNode }) => <>{children}</>,
  motion: {
    div: (props: HTMLAttributes<HTMLDivElement>) => <div {...props} />,
  },
}));

// Stub lucide-react icons
jest.mock("lucide-react", () => ({
  ChevronDown: () => <svg data-testid="chevron-down" />,
  ChevronUp: () => <svg data-testid="chevron-up" />,
}));


describe("MobileDrawer", () => {
  const fakeSummary = {
    "2023": [1, 2],
    "2021": [3],
    "2020": [4],
    "2019": [5],
  };

  beforeEach(() => {
    // Stub global.fetch to return our fake summary
    global.fetch = jest.fn().mockResolvedValue({
      json: async () => fakeSummary,
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders the newest seasons, expands weeks on click, " +
    "and only shows 'View Older Seasons' if there are older ones", async () => {
    render(<MobileDrawer isOpen={true} onClose={jest.fn()} />);

    // Wait for your useEffect and then fetch
    await waitFor(() => expect(global.fetch).toHaveBeenCalledWith("/api/recapsSummary"));

    // sortedSeasons descending order â†’ ["2023","2021","2020","2019"]
    // latestSeasons = ["2023","2021","2020"]
    // olderSeasons = ["2019"]

    // findByText query (or findBy[Whatever]) combines getByText (or [Whatever]) + waitFor
    // under the hood
    // wait for them and then assert they're in the DOM
    const season2023 = await screen.findByText("2023 Season");
    const season2021 = screen.getByText("2021 Season");
    const season2020 = screen.getByText("2020 Season");
    expect(season2023).toBeInTheDocument();
    expect(season2021).toBeInTheDocument();
    expect(season2020).toBeInTheDocument();

    // Showing another way to wait for the DOM to be ready with the elements present
    await waitFor(() => {
      expect(screen.getByText("2023 Season")).toBeInTheDocument();
    });
    expect(screen.getByText("2021 Season")).toBeInTheDocument();
    expect(screen.getByText("2020 Season")).toBeInTheDocument();

    // Asserts there is at least one older season,which makes the toggle button shows
    const toggleBtn = screen.getByText("View Older Seasons");
    expect(toggleBtn).toBeInTheDocument();

    // Expand 2023 and assert Week 1 and 2 are present
    fireEvent.click(screen.getByText("2023 Season"));
    expect(screen.getByText("Week 1")).toBeInTheDocument();
    expect(screen.getByText("Week 2")).toBeInTheDocument();

    // Toggle older
    fireEvent.click(toggleBtn);
    expect(screen.getByText("2019 Season")).toBeInTheDocument();

    // Collapse older and assert it's no longer in the DOM
    fireEvent.click(toggleBtn);
    expect(screen.queryByText("2019 Season")).toBeNull();
  });

  it("matches snapshot when open", async () => {
    const { asFragment } = render(<MobileDrawer isOpen={true} onClose={jest.fn()} />);
    // Wait for the fetch then the state to update then the render of the season links
    await screen.findByText("2023 Season");

    // After async work has been completed, Snapshot should include seasons
    expect(asFragment()).toMatchSnapshot();
  });
});
