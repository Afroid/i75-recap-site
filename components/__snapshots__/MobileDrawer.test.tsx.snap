// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileDrawer matches snapshot when open 1`] = `
<DocumentFragment>
  <div
    animate="[object Object]"
    class="fixed top-12 left-0 right-0 bottom-0 bg-black/50 z-[1000]"
    data-testid="mobile-drawer-overlay"
    exit="[object Object]"
    initial="[object Object]"
  >
    <div
      animate="[object Object]"
      class="   fixed top-12 left-0 w-60 h-[calc(100vh-3rem)]   bg-gray-50   pl-4 pr-6 py-2 z-[1001]   flex flex-col gap-6"
      data-testid="mobile-drawer"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div>
        <h2
          class="font-bold text-lg text-green-600 mb-2 underline"
          data-testid="mobile-drawer-title"
        >
          Recaps
        </h2>
        <div
          class="mb-2"
        >
          <button
            class="   flex justify-between items-center w-full   rounded-md whitespace-nowrap text-gray-900   ml-2   "
            data-testid="mobile-season-toggle-2023"
          >
            2023 Season
            <svg
              data-testid="chevron-down"
            />
          </button>
        </div>
        <div
          class="mb-2"
        >
          <button
            class="   flex justify-between items-center w-full   rounded-md whitespace-nowrap text-gray-900   ml-2   "
            data-testid="mobile-season-toggle-2021"
          >
            2021 Season
            <svg
              data-testid="chevron-down"
            />
          </button>
        </div>
        <div
          class="mb-2"
        >
          <button
            class="   flex justify-between items-center w-full   rounded-md whitespace-nowrap text-gray-900   ml-2   "
            data-testid="mobile-season-toggle-2020"
          >
            2020 Season
            <svg
              data-testid="chevron-down"
            />
          </button>
        </div>
        <div
          class="border-t border-gray-300 my-2"
        />
        <button
          class="mb-2 text-md text-green-600"
          data-testid="mobile-toggle-older-seasons"
        >
          View Older Seasons
        </button>
        <div
          class="mb-2"
        >
          <a
            class="text-md text-blue-600"
            data-testid="view-all-recaps"
            href="/recaps/viewAllRecaps"
          >
            View All Recaps
          </a>
        </div>
        <div
          class="border-t border-gray-300"
        />
      </div>
      <div
        class="flex flex-col gap-4"
      >
        <a
          class="text-lg text-gray-800"
          data-testid="nav-link-home"
          href="/"
        >
          Home
        </a>
        <a
          class="text-lg text-gray-800"
          data-testid="nav-link-about"
          href="/about"
        >
          About
        </a>
        <a
          class="text-lg text-gray-800"
          data-testid="nav-link-contact"
          href="/contact"
        >
          Contact
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
